staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/generic/controller.go

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generic
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+var _ Controller[runtime.Object] = &controller[runtime.Object]{}
+
+type controller[T runtime.Object] struct {
+	informer Informer[T]
+	queue    workqueue.RateLimitingInterface
+
+	// Returns an error if there was a transient error during reconciliation
+	// and the object should be tried again later.
+	reconciler func(namespace, name string, newObj T) error
+
+	options ControllerOptions
+}
+
+type ControllerOptions struct {
+	Name    string
+	Workers uint
+}
+
+func (c controller[T]) Informer() Informer[T] {
+	return c.informer
+}
+
+func New[T runtime.Object](
```

Comment:
wrote this generic controller since I got tired of writing the same controller boilerplate over and over

not sure best place to introduce it into k8s codebase. I figure it is low risk to keep it as a local package here that can be moved later if others find it useful?

Reply:
This looks really useful.  @andrewsykim do you have any thoughts?

Reply:
Agreed it seems super useful -- but I wonder if k8s.io/client-go (or some other repo) would be a better home for boilerplate controller code using generics. Otherwise, you're introducing new interfaces/constructors that would only be used by the CEL admission. For example, files like [cel/generic/interface.go](https://github.com/kubernetes/kubernetes/pull/112858/files#diff-9890703de1944d85d589c1b575dc431a293d187737193149d6a0f108568c7a53) and  [cel/generic/lister.go](https://github.com/kubernetes/kubernetes/pull/112858/files#diff-7f0cc04847458f416a5ef029491ae495e19b816c5ffddf7d338ec32b213f203f) would otherwise not be needed if we didn't use generics and used existing libraries.

Reply:
(I would open a separate draft PR with this implementation in k8s.io/client-go to solicit more feedback specifically on using generics for controllers as it could have a ripple effect on all the other controllers in the repo)

Reply:
Could also move it to something like `cel/internal/generic` to make it more clear that it is an implementation detail.

Reply:
If this doesn't make it into client-go, I expect that someone else will eventually duplicate the work to provide a (hopefully not inferior) solution in client-go. It would be good to at least have a rough plan. Maybe comment on https://github.com/kubernetes/kubernetes/issues/106846 and see how this would fit into the community plan for client-go generics? I don't think we should block this PR. Having the generic base here makes this code more organized and maintainable..